### ISSUE TITLE
DataTable: FilterValue EL not evaluated

### ISSUE THREAD

**Issue by mertsincan** [primefaces/primefaces#9990]
Reported By PRO User;
> We are using the data table with filtering in an area which is reloaded via Ajax.
An external operation changes the data model of the table, but the filter of the data model is not persisted in/updated from our custom data model.
For example p:column filterValue="#{ourDataModel.filterValueOfColumnXy}"
The getter and setter of the binding are not invoked on filtering, so when switching the data model, the filter of the data table is used as it was before, but not taken from out data model.

Test;
[test.zip](https://github.com/primefaces/primefaces/files/11149161/test.zip)

Here's a reproducer.

Steps to reproduce:
1. Select "Asset 1" from the left table
=> right side updates
2. Enter "A Group 2" in the filter of the "Available Groups" table
3. Select "Asset 2" from the left table
=> right side updates
=> data model is changed and the "available groups" of "Asset 2" are filtered, however with the filter value of "A Group 2" entered before.

>Our expectation is, that the expression

filterValue="#{testView.currentGroupSelectionModel.availableGroupsFilterValues['displayName']}"

>should be evaluated, when the right side is updated for the updated data model.
Also the filter value for the previously selected model should be stored - the mutator should be invoked.
Please note, that it makes no difference if the row click on "assets" is not limited via "process" or "partialSubmit" or not.

**Comment by melloware** [comment]
Can you try this...
[pf-9990.zip](https://github.com/primefaces/primefaces/files/11149731/pf-9990.zip)

I basically just added this...

```js
oncomplete="PF('wgtAvailableGroups').clearFilters();"
```

Making sure the table has a widget var

```xml
widgetVar="wgtAvailableGroups"
```

**Comment by Rapster** [comment]
> Our expectation is, that the expression filterValue="#{testView.currentGroupSelectionModel.availableGroupsFilterValues['displayName']}" should be evaluated, when the right side is updated for the updated data model.

That's the issue? Or am I supposed to see something?

If that's the issue, then I think it's because `FilterMeta` are view scoped, and gets updated only after filtering. So basically, the EL gets decoded only once. 

To see if that's solve the issue, replace `DataTable#preRender` method by this:

```java
    protected void preRender(FacesContext context, DataTable table) {
        table.setFilterByAsMap(table.initFilterBy(context));
        table.setSortByAsMap(table.initSortBy(context));
        ...
    }
```
That's something I had in mind for a while, I would like to have those metas request scoped to avoid such issues. The very first time I refactor DT, I thought having them in view scoped to gain some perf... And since we removed column from them, it's probably the total opposite :(

**Comment by mertsincan** [comment]
My suggestion was to call the reset method of dataTable in;
> 3. Select "Asset 2" from the left table

```
dt.reset(); // clear all states
```

@melloware @Rapster, nice feedback. Thanks a lot! Now I have invited the user to this ticket. I am sure we'll find the most suitable solution.

**Comment by Rapster** [comment]
@mertsincan any news on this? I could actually be wrong on this, because I can see the filterValue is decoded in `DataTable#processUpdates` (fix for #8992)

**Comment by MB4IAMS** [comment]
@melloware Thanks for the workaround. Checked it out and it clears the filter as expected from the clearFilters call - but the required behavior is to keep the filter for the selected asset when switching

@mertsincan Tried with the suggested workaround of reset() on the table as well, but this produces other problems.
1. The filter inputs are cleared and don't show the value of the "filterByAsMap"
2. The updated table first shows as empty, only after typing in the filter input again, it filters
3. Calling filter in an oncomplete-hook does not work either, as the filter inputs in the table header are not updated when the table is updated via the ajax call

So unfortunately neither approach satisfies the customer - although the clearFilters-Workaround is slightly less painful than the default behavior.

I think if the ajax update would actually update the filter inputs, we could try with the widgetVar and an oncomplete hook calling filter() again

Appreciate any other ideas...

**Comment by melloware** [comment]
Oh OK @MB4IAMS i thought you wanted to clear the filters on every change of row. So just to be clear of your expected behavior.

1. Click on Record 1 in Table 1 and then filter Table 2 (on the right).
2. Click on Record 2 in Table 1 should show all the new values in Table 2.
3. Click back on Record 1 in Table 1 and it should show the filters that were originally selected in Step 1

Is that correct?

**Comment by MB4IAMS** [comment]
@melloware yes that is correct - basically you want to keep the table filter on the groups side when switching between assets
1. Click on Record 1 in Table 1 and then filter Table 2 (on the right) 
=> Type "A Group 2"
=> only one matching "A" group should be visible
2. Click on Record 2 in Table 1 should show all the new values in Table 2.
=> Type "B Group 1"
=> only one matching "B" group should be visible
3. Click back on Record 1 in Table 1 and it should show the filters that were originally selected in Step 1
=> Filter value should show "A Group 2"
=> only one matching "A" group should be visible

That is the original approach with the value binding to filterBy being a part of the group selection model, which is stored for each asset...

**Comment by mertsincan** [comment]
@Rapster, please see the above comments. @MB4IAMS will give us information about the cases.

**Comment by Rapster** [comment]
Ok,  turns out I wasn't completely mistaken in my coment https://github.com/primefaces/primefaces/issues/9990#issuecomment-1496607234

If instead of storing filterByAsMap in the `StateHelper`, we have it as a simple attribute of `DataTable` it'll work... Quick precision from what I previously said:

> And from now on, `FilterMeta#filterValue` will be decoded **"in memory"**, but **not from** the original EL

But calling `DataTable#reset` should work, the tricky part is when should you do it? PostFilterEvent?

**Comment by sebsoftware** [comment]
Regarding this issue. If I remove

InternalPropertyKeys.filterByAsMap from the StateHelper in DataTableRender -> preRender the right side will be updated with a filtered list, filtered by the previously entered value, but the filter field will be cleared.
